contracts/upgradeable_contracts/ChaiConnector.sol

```diff
@@ -0,0 +1,269 @@
+    }
+
+    /**
+    * @dev Withdraws all invested tokens, pays remaining interest, removes chai token from contract storage
+    */
+    function removeChaiToken() external onlyOwner chaiTokenEnabled {
+        _convertChaiToDai(investedAmountInDai());
+        _payInterest();
+        delete boolStorage[CHAI_TOKEN_ENABLED];
+    }
+
+    /**
+     * @return Configured address of a receiver
+     */
+    function interestReceiver() public view returns (ERC677Receiver) {
+        return ERC677Receiver(addressStorage[INTEREST_RECEIVER]);
+    }
+
+    /**
+     * Updates interest receiver contract address
+     * @param receiver New receiver contract address
+     */
+    function setInterestReceiver(address receiver) external onlyOwner {
+        addressStorage[INTEREST_RECEIVER] = receiver;
+    }
+
+    /**
+     * @return Timestamp of last interest payment
+     */
+    function lastInterestPayment() public view returns (uint256) {
+        return uintStorage[LAST_TIME_INTEREST_PAYED];
+    }
+
+    /**
+     * @return Configured minimum interest collection period
+     */
+    function interestCollectionPeriod() public view returns (uint256) {
+        return uintStorage[INTEREST_COLLECTION_PERIOD];
+    }
+
+    /**
+     * @dev Configures minimum interest collection period
+     * @param period collection period
+     */
+    function setInterestCollectionPeriod(uint256 period) external onlyOwner {
+        uintStorage[INTEREST_COLLECTION_PERIOD] = period;
+    }
+
+    /**
+    * @dev Pays all available interest, in Dai tokens.
+    * Upgradeability owner can call this method without time restrictions,
+    * for others, the method can be called only once a specified period.
+    */
+    function payInterest() external chaiTokenEnabled {
+        // solhint-disable not-rely-on-time
+        if (
+            lastInterestPayment() + interestCollectionPeriod() < now ||
+            IUpgradeabilityOwnerStorage(this).upgradeabilityOwner() == msg.sender
+        ) {
+            uintStorage[LAST_TIME_INTEREST_PAYED] = now;
+            _payInterest();
+        }
+        // solhint-enable not-rely-on-time
+    }
+
+    /**
+    * @dev Internal function for paying all available interest, in Dai tokens
+    */
+    function _payInterest() internal {
+        require(address(interestReceiver()) != address(0));
+
+        // since investedAmountInChai() returns a ceiled value,
+        // the value of chaiBalance() - investedAmountInChai() will be floored,
+        // leading to excess remaining chai balance
+        uint256 balanceBefore = daiBalance();
+        chaiToken().exit(address(this), chaiBalance().sub(investedAmountInChai()));
+        uint256 interestInDai = daiBalance().sub(balanceBefore);
+
+        erc20token().transfer(interestReceiver(), interestInDai);
+
+        if (AddressUtils.isContract(interestReceiver())) {
+            interestReceiver().onTokenTransfer(address(this), interestInDai, "");
```

Comment:
Consider to call `onTransferToken` in more safe way similar we call in the bridgeable token:

https://github.com/poanetwork/tokenbridge-contracts/blob/efbbdfade1d2eae6a6c6f12f1a59aa936470ad6f/contracts/ERC677BridgeToken.sol#L68

It will guarantee that the call will not be reverted even if the contract receiver does not support this method call.